### PR TITLE
Move advanced IServiceCollection extensions

### PR DESCRIPTION
**Pull request by henkmollema**
Moved advanced `IServiceCollection` extensions methods used primarily by frameworks into a separate namespace: `Microsoft.Framework.DependencyInjection.Extensions`.

Moved:
- `Add(IServiceCollection, ServiceDescriptor)`
- `Add(IServiceCollection, IEnumerable<ServiceDescriptor>)`
- `TryAdd(IServiceCollection, ServiceDescriptor)`
- `TryAdd(IServiceCollection, IEnumerable<ServiceDescriptor>)`
- `TryAdd*`

`ServiceCollectionExtensions` now only consists of `AddTransient`, `AddScoped`, `AddSingleton` and `AddInstance` methods.

I'm not sure if the methods I consider as 'advanced' are the same as you guys do. Shoot if you want me to move things back and forth.

#254

### CODE REVIEW

**Comment by henkmollema** [comment]
Not sure why test fails, runs fine locally.

**Comment by pranavkm** [comment]
Tests on Travis are currently failing because of 3rd party DI tests. Those should get fixed soon.

Changes look good to me. Would you be interested in creating follow up PRs in other repos where these methods are used?

**Comment by henkmollema** [comment]
Sure!

File: src/Microsoft.Framework.DependencyInjection.Abstractions/FrameworkServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,279 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.DependencyInjection.Abstractions;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.Framework.DependencyInjection.FrameworkExtensions
```
Comment (Eilon) [inline src/Microsoft.Framework.DependencyInjection.Abstractions/FrameworkServiceCollectionExtensions.cs]: I strongly dislike this namespace name, though I can't immediately come up with a better name. Either way, please move the file into a folder that matches the namespace name.

Comment (henkmollema) [inline src/Microsoft.Framework.DependencyInjection.Abstractions/FrameworkServiceCollectionExtensions.cs]: Agree, just took the suggestion of halter73. What about just `Extensions` or `Advanced`?

Comment (Eilon) [inline src/Microsoft.Framework.DependencyInjection.Abstractions/FrameworkServiceCollectionExtensions.cs]: Maybe just `Extensions` for now? I don't like that either, but I dislike it less :smile:

Comment (henkmollema) [inline src/Microsoft.Framework.DependencyInjection.Abstractions/FrameworkServiceCollectionExtensions.cs]: @Eilon updated.

**Comment by Eilon** [comment]
Thanks, can you squash the commits so that we can merge?

**Comment by henkmollema** [comment]
@Eilon squashed and rebased! :smile:

I'll make a list a list of repos to update.

**Comment by henkmollema** [comment]
Repos which need to react:
- [ ] Mvc
- [ ] EntityFramework
- [ ] Antiforgery
- [ ] EventNotification

More..?

**Comment by HaoK** [comment]
I'm not in love with FrameworkExtensions as a namepace...  Maybe something bland like Infrastructure is better?

**Comment by henkmollema** [comment]
@HaoK I renamed it `Extensions` for now. As discussed with @Eilon.

**Comment by HaoK** [comment]
Ah I missed that thread since it was in a outdated diff... least I suggested the same name :)

**Comment by henkmollema** [comment]
Yeah :smile: 

First comment was still mentioning `FrameworkExtensions` though, corrected it now.

**Comment by halter73** [comment]
It's nice to have someone doing my work for me :smile:  :+1:

**Comment by henkmollema** [comment]
Rebased with dev, CI is green now :smile_cat:

**Comment by henkmollema** [comment]
@Eilon when are you planning to merge this? I'll have some spare time this week to create the follow-up PRs.

**Comment by pranavkm** [comment]
Sorry, I didn't see the follow up changes in other repos and wasn't sure if that was ready to go. We can get this change in today and work our way through others if they aren't already good to go.

**Comment by henkmollema** [comment]
The PRs aren't there yet. I figured we wait until this change flowed through the other components and start fixing them.

**Comment by pranavkm** [comment]
Merged.

**Comment by henkmollema** [comment]
Cool! I'll look into the follow up PRs tomorrow. :smile:

**Pull request closed without merging**